src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinRTWebSocket.cs

```diff
@@ -515,5 +549,80 @@ private void UpdateStateCloseReceived()
+            candidateCerts = candidateCerts.Find(X509FindType.FindByKeyUsage, X509KeyUsageFlags.DigitalSignature, false);
+            candidateCerts = candidateCerts.Find(X509FindType.FindByApplicationPolicy, ClientAuthenticationOID, false);
+
+            // Build a new collection with certs that have a private key. We need to do this manually because there is
+            // no X509FindType to match this criteria.
+            var eligibleCerts = new X509Certificate2Collection();
+            foreach (X509Certificate2 cert in candidateCerts)
+            {
+                if (cert.HasPrivateKey)
+                {
+                    eligibleCerts.Add(cert);
+                }
+            }
+
+            if (eligibleCerts.Count > 0)
+            {
+                return eligibleCerts[0];
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        // There are currently only two ways to convert a .NET X509Certificate2 object into a WinRT Certificate without
+        // losing its private keys, each with its own limitations:
+        //
+        // (1) Using the X509Certificate2.Export method with PKCS12/PFX to obtain a byte[] representation (including private
+        //     keys) that can then be passed into the IBuffer-based WinRT Certificate constructor. Unfortunately, the
+        //     X509Certificate2.Export operation will only succeed if the app-provided X509Certificate2 object was created
+        //     with the non-default X509KeyStorageFlags.Exportable flag.
+        //
+        // (2) Going through the certificate store. That is, retrieving the certificate represented by the X509Certificate2
+        //     object as a WinRT Certificate via WinRT CertificateStores APIs. Of course, this requires the certificate to
+        //     have been added to a certificate store in the first place.
+        //
+        // Furthermore, WinRT WebSockets only support certificates that have been added to the personal certificate store
+        // (i.e., "MY" store) due to other WinRT-specific private key limitations. With that in mind, approach (2) is the
+        // most appropriate for our needs, as it guarantees that WinRT WebSockets will be able to handle the resulting
+        // WinRT Certificate during ConnectAsync.
+        private static RTCertificate ConvertDotNetClientCertToWinRtClientCert(X509Certificate2 dotNetCertificate)
+        {
+            var query = new RTCertificateQuery
+            {
+                Thumbprint = dotNetCertificate.GetCertHash(),
+                IncludeDuplicates = false,
+                StoreName = "MY"
+            };
+
+            IReadOnlyList<RTCertificate> certificates = RTCertificateStores.FindAllAsync(query).AsTask().GetAwaiter().GetResult();
+            if (certificates.Count > 0)
+            {
+                return certificates[0];
+            }
+
+            throw new NotSupportedException(string.Format(
```

Comment:
Is NotSupportedException appropriate here?  I thought that was more of "there's nothing you can do to fix this without rebooting into a different OS/OS-version or changing from one .NET implementation to another".  Not "please add your cert to the CU\My store"